presto-spi/src/main/java/com/facebook/presto/spi/block/BlockUtil.java

```diff
@@ -30,11 +27,27 @@ private BlockUtil()
     {
     }
 
-    static void checkValidPositions(List<Integer> positions, int positionCount)
+    static void checkValidPositions(int[] positions, int offset, int length, int positionCount)
     {
-        Set<Integer> invalidPositions = positions.stream().filter(position -> position >= positionCount).collect(toSet());
-        if (!invalidPositions.isEmpty()) {
-            throw new IllegalArgumentException("Invalid positions " + invalidPositions + " in block with " + positionCount + " positions");
+        checkValidPositionsArray(positions, offset, length);
+
+        for (int i = 0; i < length; ++i) {
+            int position = positions[offset + i];
+            if (position > positionCount) {
+                throw new IllegalArgumentException("Invalid position " + position + " in block with " + positionCount + " positions");
+            }
+        }
+    }
+
+    static void checkValidPositionsArray(int[] positions, int offset, int length)
+    {
+        requireNonNull(positions, "positions array is null");
+        if (offset < 0 || offset > positions.length) {
```

Comment:
nit: You could merge the conditions into single if and drop the `offset > positions.length` part.

Reply:
if you merge this conditions, I recommend using exc msg like in #9391